### PR TITLE
fix(public-stats): skip malformed v1 dimension entries instead of 500ing

### PR DESCRIPTION
**Pull request by Zingzy**
## The bug (empirically hit on beta)

`PublicStatsService._v1_dimension_rows` adapts the legacy v1 analytics dicts (`browser`, `os_name`, `country`, `referrer` — shape `{value: entry}`). It handled entry as `{"counts": n, "ips": [...]}` and, defensively, as a bare int-able scalar. Any other shape fell into `int(entry or 0)` and raised. Real beta traceback:

```
{"error": "int() argument must be a string, a bytes-like object or a real number, not 'list'",
 "error_type": "TypeError", "path": "/api/v1/public/stats/spring", "event": "unhandled_exception"}
```

**Blast radius:** the exception is unhandled, so ONE malformed sub-entry anywhere in a v1 doc makes that link's public stats page 500 on **every request, forever** — post-cutover visitors get the composed 5xx apology page for stats that are otherwise perfectly renderable. v1 data spans four years of Flask-era schema drift; the defensive scalar branch exists *because* shapes vary — it just didn't go far enough.

## The fix — skip, don't sink

- An entry whose count can't be parsed (bare list — the live repro — dict with junk `counts`, non-numeric scalar) is **dropped**; every parseable entry in the dimension still renders.
- No guessed semantics: a list is not `len()`-counted into a click count — skipping is honest, inventing numbers isn't.
- One `v1_dimension_entries_skipped` info log per dimension naming the skipped values — not one log per row, so a crusty doc can't spam.
- Bonus hardening in the same function: a parseable count with an unrecognisable `ips` container (e.g. a bare string) keeps the honest click count and reports zero uniques instead of crashing/garbage.
- Happy paths are byte-identical.

## Time-series loops (inspected, deliberately unchanged)

The `int(count or 0)` loops over `counter`/`unique_counter`/`bots` look like the same hazard but are **unreachable with non-scalar values**: those fields are typed `dict[str, int]` on `LegacyUrlDoc`, so pydantic coerces scalar drift (`"7"` → 7, `2.0` → 2) and rejects non-scalars at `from_mongo` — before the synthesis loops run (verified empirically). The dimension maps are `dict[str, Any]`, which is exactly why drift reaches `_v1_dimension_rows` and nowhere else. A hypothetical list-valued counter bucket would fail at model validation — a layer shared with the redirect path (`find_by_id`), where such docs would already be failing loudly; no evidence any exist (counters are written via `$inc`, which can only produce numbers). Loosening the shared model is a bigger call than this fix warrants.

## Tests

+3 (2193 → 2196), all green; `ruff check` + `ruff format` clean.

- `test_v1_malformed_dimension_entries_are_skipped_not_500` — live-repro list shape, junk `counts`, non-numeric scalar, each next to a clean sibling → 200, clean entries render, malformed absent.
- `test_v1_every_entry_malformed_yields_empty_dimension_not_500` — worst case degrades to an empty dimension, not a 500.
- `test_v1_non_list_ips_container_keeps_count_drops_uniques` — pins the count-stands/uniques-drop behavior.
- Comment in the test file pins *why* there's no malformed time-series test (unreachable via pydantic typing).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by Sourcery

Harden v1 public stats dimension processing to tolerate drifted legacy analytics data by skipping unparseable entries, preserving valid metrics, and adding targeted logging and tests.

Bug Fixes:
- Prevent malformed v1 analytics dimension entries from causing 500 errors by skipping unparseable entries and hardening IP container handling.

Enhancements:
- Log skipped v1 analytics dimension entries once per dimension for better visibility into legacy schema drift.

Tests:
- Add integration tests covering malformed v1 dimension entries, all- malformed dimensions yielding empty results, and non-list IP containers keeping counts but dropping uniques.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved resilience when displaying legacy statistics containing malformed dimension data.
  * Malformed or non-integer legacy counter values are now skipped instead of causing server errors.
  * If the uniques payload is malformed, total click counts are still rendered while uniques are safely omitted or set to zero as appropriate.
  * Added coverage to ensure the endpoint stays HTTP 200 and returns empty/sanitized dimension results when all dimension entries are invalid.

### CODE REVIEW

**Review by Zingzy — COMMENTED** [review]
## Review: skip malformed v1 dimension entries

**Mergeable, no blockers.** A tight, correct resilience fix for the `PublicStatsService` from #241 — a real beta 500 (a bare-list dimension entry hitting `int()`) turned into skip-don't-sink. I traced every branch and verified the scope claim against the model (`schemas/models/url.py:132-138`); CI green 3.10–3.14 (2196).

**Why it's right:**
- Every unparseable shape the traceback and drift produce is handled: `counts: ["huh"]` → `int([...])` TypeErrors → caught → skip; a bare `["1.1.1.1"]` (the live repro) → `else` branch `int(list)` → skip; `"not-a-number"` → ValueError → skip. Parseable siblings still render, and numeric strings (`"12"`) still parse — no happy-path regression.
- **The `ips` isinstance guard is a genuine catch, not just defensiveness:** the old `set().update(ips)` with a *string* `ips="1.1.1.1"` would have iterated it and added `'1'`, `'.'`, … as individual "unique IPs" — silent garbage, not a crash. Zeroing non-list containers turns that into an honest 0.
- **Skip, don't invent:** a list isn't `len()`-counted into clicks. Dropping is honest; fabricating a number would poison the analytics the page exists to show.
- **Correctly scoped restraint:** leaving the `int(count or 0)` time-series loops alone is *justified*, not a gap — `counter`/`unique_counter`/`bots` are `dict[str, int]`, so non-scalar drift fails at `from_mongo` (shared with the redirect's `find_by_id`, where it'd already be loud), and `$inc` only writes numbers; the dimension maps are `dict[str, Any]`, which is exactly why drift reaches `_v1_dimension_rows` and nowhere else. Loosening a shared model for a hypothetical would be a bigger, riskier change than this bug warrants. The test-file NOTE pinning *why* there's no time-series test is the right artifact.

**On the Sourcery feedback — three minor observability nits, none blocking:**
- **Info log fires per drifted request** — the most valid, but this is a low-volume endpoint (public stats, not the hot path), it's one log per *dimension* not per row, and lowering to `debug` would sacrifice the visibility that's the point (you *want* to know which legacy links are drifted, once). Leave it, or sample if it ever gets noisy.
- **`values=skipped` could grow large** — bounded by a dimension's distinct-value cardinality, but a popular link with hundreds of drifted referrers is conceivable. If you touch anything, capping to the first ~20 (`+N more`) is the one tidy micro-nit worth entertaining. Optional.
- **Log the bad `ips` type** — the non-list-ips path isn't a skip (count still renders), so it's a silent uniques-drop; a log line would help an operator distinguish it, but risks more noise for a rare case. Optional; the test already documents the behavior.

**Blast radius / merge safety:** v1 public-stats read path only, v1 links only — strictly an upgrade (previously-500ing pages now render minus the junk rows; happy paths byte-identical). No flag; live on the next release like the rest of the public-stats surface (#241), and correctness is the gate, which holds.

**What's good:** precise repro→fix discipline, the honesty of skip-over-invent, the `ips` guard catching the char-level garbage bug, the scope restraint with a written rationale, and three well-chosen tests (repro + all-malformed + non-list-ips) plus the NOTE documenting the deliberately-untested path.

File: services/public_stats_service.py [hunk 1]
```diff
@@ -403,20 +403,47 @@ def _v1_dimension_rows(
...
-                count = int(entry or 0)
+                try:
+                    count = int(entry or 0)
+                except (TypeError, ValueError):
+                    skipped.append(raw_value)
+                    continue
                 ips = []
             value = transform(raw_value) if transform else raw_value
             clicks[value] = clicks.get(value, 0) + count
             unique_ips.setdefault(value, set()).update(ips)
```
[the remark this answers is not in the document]
Comment (Zingzy) [reply services/public_stats_service.py:439]: Real catch: nested-list elements passed the container guard and TypeError'd `set.update` outside both try/excepts, the same 500 class this branch closes. Fixed in 48115a0: non-str elements are dropped before the union (v1 only ever wrote plain strings), count stands, string siblings still count as uniques. Pinned by `test_v1_unhashable_ips_elements_keep_count_and_clean_uniques` with the exact repro shape.
